### PR TITLE
statistics: reduce unnecessary cpu time

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Issue Number: Close https://github.com/tikv/pd/issues/5691

### What is changed and how does it work?

1. cache `Gauge.WithLabelValues`
2. reduce log zap
3. improve `collectExpiredItems` and `isOldColdPeer` and `justTransferLeader`

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

reduce 16% (from 110s to 92s)

Before:
image

After:
image

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/statistics/hot_peer_cache.go [hunk 1]
```diff
@@ -157,12 +171,14 @@ func (f *hotPeerCache) collectPeerMetrics(loads []float64, interval uint64) {
 func (f *hotPeerCache) collectExpiredItems(region *core.RegionInfo) []*HotPeerStat {
 	regionID := region.GetID()
 	items := make([]*HotPeerStat, 0)
-	for _, storeID := range f.getAllStoreIDs(region) {
-		if region.GetStorePeer(storeID) == nil {
-			item := f.getOldHotPeerStat(regionID, storeID)
-			if item != nil {
-				item.actionType = Remove
-				items = append(items, item)
+	if ids, ok := f.storesOfRegion[regionID]; ok {
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:174]: length of `f.storesOfRegion` is less or equal than `f.getAllStoreIDs`

File: server/statistics/hot_peer_cache.go [hunk 2]
```diff
@@ -294,51 +308,70 @@ func (f *hotPeerCache) getOldHotPeerStat(regionID, storeID uint64) *HotPeerStat
...
+		for i := range t.rates {
+			t.rates[i] = math.Max(tn.GetTopNMin(i).(*HotPeerStat).GetLoad(i)*HotThresholdRatio, t.rates[i])
+		}
 	}
 	return t.rates
 }
 
 // gets the storeIDs, including old region and new region
 func (f *hotPeerCache) getAllStoreIDs(region *core.RegionInfo) []uint64 {
-	storeIDs := make(map[uint64]struct{})
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:322]: the length of peers is enough small, map consume unnecessary

**Review by nolouch — APPROVED** [review]
lgtm

**Comment by nolouch** [comment]
ptal @rleungx

File: server/statistics/hot_peer.go [hunk 3]
```diff
@@ -126,18 +127,20 @@ func (stat *HotPeerStat) Less(dim int, than TopNItem) bool {
...
-		zap.Float64s("loads", stat.GetLoads()),
-		zap.Float64s("loads-instant", stat.Loads),
-		zap.Int("hot-degree", stat.HotDegree),
-		zap.Int("hot-anti-count", stat.AntiCount),
-		zap.Duration("sum-interval", stat.getIntervalSum()),
-		zap.Bool("allow-inherited", stat.allowInherited),
-		zap.String("action-type", stat.actionType.String()),
-		zap.Time("last-transfer-leader-time", stat.lastTransferLeaderTime))
+func (stat *HotPeerStat) Log(str string) {
+	if log.GetLevel() <= zap.DebugLevel {
```
Comment (rleungx) [inline server/statistics/hot_peer.go:131]: Why not just use `log.Debug`?

File: server/statistics/hot_peer_cache_test.go [hunk 4]
```diff
@@ -706,6 +703,40 @@ func TestHotPeerCacheTopN(t *testing.T) {
 	re.Equal(3200.0, cache.calcHotThresholds(1)[KeyDim])
 }
 
+func TestHotPeerCacheTopNLen(t *testing.T) {
```
Comment (rleungx) [inline server/statistics/hot_peer_cache_test.go:706]: This test is almost the same as `TestHotPeerCacheTopNThreshold`

File: server/statistics/hot_peer.go (hunk 3 above)
Comment (lhy1024) [reply server/statistics/hot_peer.go:131]: it will cost unnecessary cpu for zap

**Review by rleungx — APPROVED** [review]
The rest LGTM.

**Pull request merged**